### DISCUSSION TITLE
Section 4 NFT : I can't find the solution

### DISCUSSION THREAD

**Discussion by clement-touzet** [Cyfrin/security-and-auditing-full-course-s23#120] [Q&A]
I'm trying to solve the exercice in the section 4, but I can't find what is wrong with my smart contrat. 

Can someone show me their solution or explain to me what should I do to complete this exercice ?

This is my code that doesn't work : 

```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.18;

interface S4 {
    function solveChallenge(uint256 guess, string memory yourTwitterHandle) external;
}

contract SolveS4 {
    S4 s4Contract;
    constructor() {
        s4Contract = S4(0xf988Ebf9D801F4D3595592490D7fF029E438deCa);
    }

    function solveChallenge() public {
        s4Contract.solveChallenge(0, "");
    }

    function owner() view public returns (uint256,address) {
        return (type(uint256).max, address(this));
    }

    function go() public  {
        // reentrancy part
        uint guess = uint256(keccak256(abi.encodePacked(address(this), block.prevrandao, block.timestamp))) % 1_000_000;
        s4Contract.solveChallenge(guess, "clemmos_xyz");
    }
}
```

**Comment by alymurtazamemon** [comment]
@clement-touzet After your question I tried this challenge and this code is working in Remix.

```solidity
// SPDX-License-Identifier: MIT

pragma solidity 0.8.20;

import "hardhat/console.sol";

contract A {
    B b;

    constructor(address _address) {
        b = B(_address);
    }

    function solve() public {
        b.solveChallenge(0);
    }
       
    function owner() public view returns(address) {
        return address(this);
    }

    function go() public {
        uint256 guess =
                uint256(keccak256(abi.encodePacked(address(this), block.prevrandao, block.timestamp))) % 1_000_000;

        b.solveChallenge(guess);
    }
}

contract B {
    error S4__BadReturn();
    error S4__BadOwner();
    error S4__BadGuess();

    uint256 myVal = 0;

    function solveChallenge(uint256 guess) external {
        (bool success, bytes memory returnData) = msg.sender.staticcall(abi.encodeWithSignature("owner()"));
        
        address ownerAddress;
        assembly {
            ownerAddress := mload(add(returnData, 32))
        }
        console.log(ownerAddress);

        if (!success || ownerAddress != msg.sender) {
            revert S4__BadOwner();
        }
        console.log("condition 1 clear");

        if (myVal == 1) {
            uint256 rng = uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 1_000_000;
            if (rng != guess) {
                revert S4__BadGuess();
            }
            console.log("successed!");
        } else {
            myVal = 1;
            (bool succ,) = msg.sender.call(abi.encodeWithSignature("go()"));
            if (!succ) {
                revert S4__BadReturn();
            }
        }
        myVal = 0;
    }
}
```
<img width="1066" alt="Screenshot 2024-02-16 at 6 19 08 PM" src="https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/56123405/d306bea4-c2b5-48c2-9bcc-1d5003921320">

But this is failing on Testnet, I am figuring out this until then you can also check.

**Reply by clement-touzet** [reply]
@alymurtazamemon 
First, thanks for your solution, I realised my `owner()` function was not working returning the right result.

I still can't find what's is wrong, so I guess it's the `_updateAndRewardSolver(yourTwitterHandle);` line which revert the transaction, as it is the only one not in your code.

**Reply by alymurtazamemon** [reply]
@clement-touzet You welcome!

Yes, I am also figuring out from the movement you asked the question but really struggling to find the reason why it is not working on Testnet. `_updateAndRewardSolver ` function could be the reason but I am sure the designer of this challenge must have tried minting the NFT before giving the challenge so I suppose I am doing something wrong. But leaving it here and will see it after completing the course.

**Comment by npa528** [comment]
@clement-touzet Part of the solution is to clearly define what the problem is :)
If I hadn't solve the challenge I wouldn't really understand where exactly you have stuck.

I will **not** give any SPOILERS (just some hints) so if someone wants to solve this on his/her own should stop reading here!!!


-------------------------------------------------------------------------------------
*************************************** HINTS **************************************
-------------------------------------------------------------------------------------
To give you both a hint, your code is correct but incomplete.

Answer these hints below to help you understand:
1) what exactly is the reward you are waiting after you solve challenge (in terms of ERC_XXX) ? 
2) Is your contract ready to **receive** this reward? 
3) Can you make it "ready"?

PS. Above hints will help you in ALL NFT challenges, not only 4

**Reply by EngrPips** [reply]
Insightful, i mean you make it clear without actually directly saying the answer. impressive @npa528

**Comment by alymurtazamemon** [comment]
@clement-touzet I stopped reading after the Spoiler alert!

I know there was some assembly code where I am still weak so I have to solve these challenges. Waiting for the second part of the course so after it I will try to solve these. Otherwise, I will read your hint part!
